### PR TITLE
Adds the ability to use the web api with non-admin users

### LINKED ISSUE #1333: [Feature] Manage / Create API key + Manage IP access API
### In a few words, please describe the issue that you're experiencing:
Add options to manage API keys (Add / Revoke)
Manage IP access to API

+ Votes for adding certain access rights
Example:
1. Manage 1 account (Example user1)
2. Manage users account (Example user1, user2, ect...)
3. Manage Admin Settings 

Allow API access only to certain IP's

### PR DESCRIPTION
**Pull request by jhmaverick**
* Adds option to define a user for the key and the scripts that can be used;
* List of scripts that can be used by non-admin users with "user" argument position for validation;
* Keys generated for admin can still have access to all scripts but it is possible to limit which ones can be used;
* Keys in the old format will run as admin and have access to all scripts.

### CODE REVIEW

**Comment by jlguerrerosanz** [comment]
Great addition. Thank you!

**Comment by jlguerrerosanz** [comment]
I would love to know how to properly use this new feature with the dns-cluster user.

Maybe an example in the script docs.

File: bin/v-generate-api-key [hunk 1]
```diff
@@ -36,14 +57,20 @@ if [ ! -d "$HESTIA/data/keys/" ]; then
...
 fi
 
-touch $HESTIA/data/keys/$HASH
-echo "$HASH"
+echo "USER='$user'" >$HESTIA/data/keys/$HASH
+echo "SCRIPTS='$scripts'" >>$HESTIA/data/keys/$HASH
+echo "TIME='$time'" >>$HESTIA/data/keys/$HASH
+echo "DATE='$date'" >>$HESTIA/data/keys/$HASH
+
+echo "USER:    $user"
```
Comment (jaapmarcus) [inline bin/v-generate-api-key]: We should add an option to supply an argument that allows output as json for Web UI

File: bin/v-list-api-key [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    echo "TIME:       $TIME"
+    echo "DATE:       $DATE"
+}
+
+#----------------------------------------------------------#
+#                    Verifications                         #
+#----------------------------------------------------------#
+
+check_args '1' "$#" 'KEY [FORMAT]'
+is_format_valid 'key'
```
Comment (jaapmarcus) [inline bin/v-list-api-key]: .. will also be considered a valid key with is_format_valid 'key'

Validate with v-check-api-key to make sure it it valid 

 is_format_valid 'key' is used to valid config keys in user.conf for example...

**Review by jaapmarcus — CHANGES_REQUESTED** [review]
It looks fine it would be nice to allow the use of presets:

For example when using DNS cluster set preset "as"  dns-cluster and you don't have to list the fulll list you need.

Other examples:
phpmyadmin-sso
purge-nginx-cache


whmcs / billing sytem.. And so on

**Comment by jaapmarcus** [comment]
> I would love to know how to properly use this new feature with the dns-cluster user.
> 
> Maybe an example in the script docs.

At some point we need to update the docs anyway

File: bin/v-generate-api-key (hunk 1 above)
Comment (jhmaverick) [reply bin/v-generate-api-key]: I added a third argument in the script to make it possible to change the output

File: bin/v-list-api-key (hunk 2 above)
Comment (jhmaverick) [reply bin/v-list-api-key]: As this script and the "v-check-api-key" serve similar purposes, I removed the "v-list-api-key" and added the return for the v-check-api-key"

**Comment by jhmaverick** [comment]
I made some changes to simplify using the functionality and added the option for script groups in the API. The way the script obtains the position of the "USER" parameter in scripts has also been changed.

Initially I just added 2 groups "mail-accounts" and "purge-nginx-cache". I didn't add a group for "dns-cluster" because I don't know exactly which scripts are ideal for this functionality and I also noticed that some of his scripts don't have the "USER" argument to limit actions only to the defined user.

File: bin/v-list-api-key (hunk 2 above)
Comment (jaapmarcus) [reply bin/v-list-api-key]: 👍🏼

**Comment by jaapmarcus** [comment]
Merged this PR in a different branch + Some other changes Quick installers 

#2503

**Comment by jaapmarcus** [comment]
Sorry the wrong one :)

**Comment by jaapmarcus** [comment]
Did. some testing and it works fine:

Some small remarks:
Because we overwrite main.sh each update users can't update the "group" list
Some "keys" benefit even from restricting for the admin user:

https://github.com/jaapmarcus/hestiacp/tree/feature/nonadminapi-changes

Should solve this issue.

- Only thing "left" is log any kind of "faulty" usage as they are currently only "returned" with an exit code but not logged on the server side
- Make v-list-api-keys with a list of the current active "keys" 
- UI to view it in admin panel and allow admin create the keys...
-

**Comment by jhmaverick** [comment]
I liked the folder for groups, it really got better.

I think the first 2 items will be simple to do. Maybe the interface part takes more time. But I'll see what I can do.

**Comment by jaapmarcus** [comment]
Fist two items should no issues... And relative easy...

"My" changes haven't been merged into this branch feel free to do so...

**Comment by jhmaverick** [comment]
I found some problems when listing the keys, as they were composed of just one hash, the key would need to be exposed on the panel to display or even to control the delete and display events... So I thought it best to create one new system and leave the old one as a legacy.

#2535 Here is the version with the new commands and management interface

**Comment by jaapmarcus** [comment]
https://github.com/hestiacp/hestiacp/pull/2535 

Will close this one

**Pull request closed without merging**